torch/autograd/gradcheck.py

```diff
@@ -965,17 +965,22 @@ def wrapped_fn(*inputs):
 
     return apply_to_c_outs(fn, torch.real), apply_to_c_outs(fn, torch.imag)
 
-def _real_and_imag_input(fn, complex_inp_indices):
-    # returns new functions that take real inputs instead of complex inputs and compute fn(x + 0 * 1j)
-    # and f(x * 1j).
+def _real_and_imag_input(fn, complex_inp_indices, tupled_inputs):
+    # returns new functions that take real inputs instead of complex inputs as
+    # (x, y) -> fn(x + y * 1j). And it computes: inp -> fn(inp + y * 1j) and inp -> fn(x + inp * 1j).
+    # In each case, the other part is considered constant.
+    # We do not use 0 for the constant here to make sure we always call the user function with a valid input.
     def apply_to_c_inps(fn, fn_to_apply):
         def wrapped_fn(*inputs):
             new_inputs = list(inputs)
             for should_be_complex in complex_inp_indices:
-                new_inputs[should_be_complex] = fn_to_apply(new_inputs[should_be_complex])
+                new_inputs[should_be_complex] = fn_to_apply(new_inputs[should_be_complex],
+                                                            tupled_inputs[should_be_complex])
             return _as_tuple(fn(*new_inputs))
         return wrapped_fn
-    return apply_to_c_inps(fn, lambda x: x + 0 * 1j), apply_to_c_inps(fn, lambda x: x * 1j)
+    real_fn = apply_to_c_inps(fn, lambda inp, orig: inp + orig.imag * 1j)
```

Comment:
what's `inp` and `orig`? what's the effective change here? why did this fix the invalid input generation issue for the linalg tests?

Reply:
I think this fixes it because some linalg operators are non-differentiable for purely imaginary/pure real inputs, and this change makes it so that we effectively run the operator on the inputs defined in the OpInfo, as intended

I agree with @anjali411 that this naming could be better (or a comment could suffice?) about inp being a pure real quantity, in both the `real_fn` and `imag_fn` cases. And if my understand is correct, another comment about how in this wrapper function we'd like to reconstruct the original complex input while holding either the real/imag component constant, as to satisfy the constraint that inputs for complex forward AD.

Reply:
Could you clarify what you mean by
>we'd like to reconstruct the original complex input while holding either the real/imag component constant, as to satisfy the constraint that inputs for complex forward AD should be real.

I do not understand why this should be the case, or where would this limitation come from.